core/shared/src/main/scala/org/http4s/multipart/PartReceiver.scala

```diff
@@ -0,0 +1,206 @@
+    * passing `strict = false` to the underlying `decode` method.
+    *
+    * This method passes `strict = false` to the `decoder.decode` method.
+    *
+    * @param decoder The delegate EntityDecoder
+    */
+  def decode[F[_], A](implicit decoder: EntityDecoder[F, A]): PartReceiver[F, A] =
+    part => Resource.eval(decoder.decode(part, strict = false).value)
+
+  /** Creates a PartReceiver that delegates to an implicitly-resolved `EntityDecoder` to decode the part body,
+    * passing `strict = true` to the underlying `decode` method.
+    *
+    * @param decoder The delegate EntityDecoder
+    */
+  def decodeStrict[F[_], A](implicit decoder: EntityDecoder[F, A]): PartReceiver[F, A] =
+    part => Resource.eval(decoder.decode(part, strict = true).value)
+
+  /** Creates a PartReceiver that loads the part's body into an in-memory buffer, up until
+    * a specified maximum size, at which point it will instead write the body to a temp file.
+    *
+    * The temp file created by this decoder (if any) will be released with the Resource returned
+    * by this receiver's `receive` method.
+    *
+    * @param maxSizeBeforeFile Maximum number of bytes to be buffered in memory for each part
+    *                          received by the returned receiver. When the in-memory buffer
+    *                          would exceed this limit, the buffer is dumped to a temp file
+    *                          and all subsequent data pulled from the part body will be
+    *                          appended to that file instead of accumulating the memory buffer.
+    * @param chunkSize The chunk size used when reading data back from a temporary file created by this receiver
+    * @return A PartReceiver which dynamically decides whether to buffer the part's body in memory or in a file
+    */
+  def toMixedBuffer[F[_]](
+      maxSizeBeforeFile: Int,
+      chunkSize: Int = 8192,
+  )(implicit
+      files: Files[F],
+      mt: MonadThrow[F],
+      c: Compiler[F, F],
+  ): PartReceiver[F, Stream[F, Byte]] =
+    part => readToBuffer[F](part.body, maxSizeBeforeFile, chunkSize).map(Right(_))
+
+  private def limitPartSize[F[_]](
+      maxPartSizeBytes: Long
+  )(implicit F: ApplicativeError[F, Throwable]): Pipe[F, Byte, Byte] = {
+    def go(s: Stream[F, Byte], accumSize: Long): Pull[F, Byte, Unit] = s.pull.uncons.flatMap {
+      case Some((chunk, tail)) =>
+        val newSize = accumSize + chunk.size
+        if (newSize <= maxPartSizeBytes) Pull.output(chunk) >> go(tail, newSize)
+        else Pull.raiseError[F](InvalidMessageBodyFailure("Part body exceeds maximum length"))
+      case None =>
+        Pull.done
+    }
+
+    go(_, 0L).stream
+  }
+
+  private def readToBuffer[F[_]: Files: MonadThrow](
+      input: Stream[F, Byte],
+      maxSizeBeforeFile: Int,
+      chunkSize: Int,
+  )(implicit c: Compiler[F, F]): Resource[F, Stream[F, Byte]] = {
```

Comment:
I apologize if I am missing something, but I cannot figure this out: what it the point of creating a `Resource` of `Stream` here? The `Stream` type is already a scoped thing, it should not require additional scope managing. Moreover, I cannot figure out where the `Resource` created would be managing any resource per se: looks like it is either `Resource.pure` or `Resource.suspend`. Neither of them introduce any new scope for further management. Moreover, the latter `Resource.suspend` does not seem necessary at all in this context.

Reply:
Ah, ok, I see now – it is used for `Files[F].tempFile`. Sorry, missed it out.
Anyway, it does not seem necessary to enclose `Stream` into `Resource`, something like this can be used instead:
```
Stream.reasource(Files[F].tempFile(...)).flatMap { path =>
  Files[F].readAll(path)
}
```

Reply:
Responding from my phone; apologies for any errors...

`readToBuffer` is my attempt to represent the existing functionality of `parseToPartsSupervisedFile` in terms of the new `PartReceiver` trait. This method is a private implementation detail of `toMixedBuffer`, defined a few lines up from here. 

The intention here is that allocation of the resource is what causes the part body to be collected to a buffer (which is either an in-memory chunk a temp file), and the returned stream acts as a facade for that buffer while it is still valid. Reading that steam multiple times should not have to re-create the temp file, so I don't think I can Construct it the way you suggest.